hibernate.hbm2ddl.auto=update can cause `Executing queries is not allowed for DDL batches.` error

Starting Hibernate with `hibernate.hbm2ddl.auto=update` with an existing database that contains foreign keys can cause the error listed below. This happens because Hibernate tries to extract additional metadata to check which columns are present on a table using a non-standard query (that is; it does not use one of the standard JDBC `DatabaseMetaData` methods). This is not allowed inside a DDL batch.

The query that is executed by Hibernate is defined here: https://github.com/hibernate/hibernate-orm/blob/37d3b66ce411656c305910cff9d9867d8716c965/hibernate-core/src/main/java/org/hibernate/tool/schema/extract/internal/InformationExtractorJdbcDatabaseMetaDataImpl.java#L148

```
2022-10-31 18:03:53,253 WARN  [scheduling-1] - SQL Error: 9, SQLState: null
2022-10-31 18:03:53,253 ERROR [scheduling-1] - FAILED_PRECONDITION: Executing queries is not allowed for DDL batches.
2022-10-31 18:03:53,258 ERROR [main] - Unexpected exception
javax.persistence.PersistenceException: [PersistenceUnit: default] Unable to build Hibernate SessionFactory; nested exception is org.hibernate.exception.GenericJDBCException: Error accessing column metadata: <redacted>
	at org.springframework.orm.jpa.AbstractEntityManagerFactoryBean.buildNativeEntityManagerFactory(AbstractEntityManagerFactoryBean.java:403) ~[spring-orm-5.2.8.RELEASE.jar:5.2.8.RELEASE]
	at java.util.concurrent.FutureTask.run(FutureTask.java:266) ~[?:1.8.0_181]
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:180) ~[?:1.8.0_181]
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:293) ~[?:1.8.0_181]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) ~[?:1.8.0_181]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) ~[?:1.8.0_181]
	at java.lang.Thread.run(Thread.java:748) ~[?:1.8.0_181]
Caused by: org.hibernate.exception.GenericJDBCException: Error accessing column metadata: <redacted>
	at org.hibernate.exception.internal.StandardSQLExceptionConverter.convert(StandardSQLExceptionConverter.java:42) ~[hibernate-core-5.6.12.Final.jar:5.6.12.Final]
	at org.hibernate.engine.jdbc.spi.SqlExceptionHelper.convert(SqlExceptionHelper.java:113) ~[hibernate-core-5.6.12.Final.jar:5.6.12.Final]
	at org.hibernate.engine.jdbc.spi.SqlExceptionHelper.convert(SqlExceptionHelper.java:99) ~[hibernate-core-5.6.12.Final.jar:5.6.12.Final]
	at org.hibernate.tool.schema.extract.internal.AbstractInformationExtractorImpl.convertSQLException(AbstractInformationExtractorImpl.java:118) ~[hibernate-core-5.6.12.Final.jar:?]
	at org.hibernate.tool.schema.extract.internal.InformationExtractorJdbcDatabaseMetaDataImpl.addColumns(InformationExtractorJdbcDatabaseMetaDataImpl.java:172) ~[hibernate-core-5.6.12.Final.jar:?]
	at org.hibernate.tool.schema.extract.internal.AbstractInformationExtractorImpl.extractTableInformation(AbstractInformationExtractorImpl.java:844) ~[hibernate-core-5.6.12.Final.jar:?]
	at org.hibernate.tool.schema.extract.internal.AbstractInformationExtractorImpl.lambda$getTable$2(AbstractInformationExtractorImpl.java:399) ~[hibernate-core-5.6.12.Final.jar:?]
	at org.hibernate.tool.schema.extract.internal.InformationExtractorJdbcDatabaseMetaDataImpl.processTableResultSet(InformationExtractorJdbcDatabaseMetaDataImpl.java:69) ~[hibernate-core-5.6.12.Final.jar:?]
	at org.hibernate.tool.schema.extract.internal.AbstractInformationExtractorImpl.getTable(AbstractInformationExtractorImpl.java:394) ~[hibernate-core-5.6.12.Final.jar:?]
	at org.hibernate.tool.schema.extract.internal.DatabaseInformationImpl.getTableInformation(DatabaseInformationImpl.java:114) ~[gps.core.jar:?]
	at org.hibernate.tool.schema.extract.internal.DatabaseInformationImpl.locateTableInformation(DatabaseInformationImpl.java:152) ~[gps.core.jar:?]
	at org.hibernate.tool.schema.extract.internal.AbstractInformationExtractorImpl.lambda$getForeignKeys$9(AbstractInformationExtractorImpl.java:1298) ~[hibernate-core-5.6.12.Final.jar:?]
	at org.hibernate.tool.schema.extract.internal.InformationExtractorJdbcDatabaseMetaDataImpl.processImportedKeysResultSet(InformationExtractorJdbcDatabaseMetaDataImpl.java:132) ~[hibernate-core-5.6.12.Final.jar:?]
	at org.hibernate.tool.schema.extract.internal.AbstractInformationExtractorImpl.getForeignKeys(AbstractInformationExtractorImpl.java:1277) ~[hibernate-core-5.6.12.Final.jar:?]
	at org.hibernate.tool.schema.extract.internal.TableInformationImpl.foreignKeys(TableInformationImpl.java:97) ~[hibernate-core-5.6.12.Final.jar:?]
	at org.hibernate.tool.schema.extract.internal.TableInformationImpl.getForeignKeys(TableInformationImpl.java:91) ~[hibernate-core-5.6.12.Final.jar:?]
	at org.hibernate.tool.schema.internal.AbstractSchemaMigrator.equivalentForeignKeyExistsInDatabase(AbstractSchemaMigrator.java:500) ~[hibernate-core-5.6.12.Final.jar:5.6.12.Final]
	at org.hibernate.tool.schema.internal.AbstractSchemaMigrator.checkForExistingForeignKey(AbstractSchemaMigrator.java:485) ~[hibernate-core-5.6.12.Final.jar:5.6.12.Final]
	at org.hibernate.tool.schema.internal.AbstractSchemaMigrator.applyForeignKeys(AbstractSchemaMigrator.java:442) ~[hibernate-core-5.6.12.Final.jar:5.6.12.Final]
	at org.hibernate.tool.schema.internal.AbstractSchemaMigrator.performMigration(AbstractSchemaMigrator.java:263) ~[hibernate-core-5.6.12.Final.jar:5.6.12.Final]
	at org.hibernate.tool.schema.internal.AbstractSchemaMigrator.doMigration(AbstractSchemaMigrator.java:123) ~[hibernate-core-5.6.12.Final.jar:5.6.12.Final]
	at com.google.cloud.spanner.hibernate.schema.SpannerSchemaMigrator.doMigration(SpannerSchemaMigrator.java:57) ~[google-cloud-spanner-hibernate-dialect-1.5.7-SNAPSHOT.jar:1.5.7-SNAPSHOT]
	at org.hibernate.tool.schema.spi.SchemaManagementToolCoordinator.performDatabaseAction(SchemaManagementToolCoordinator.java:196) ~[hibernate-core-5.6.12.Final.jar:5.6.12.Final]
	at org.hibernate.tool.schema.spi.SchemaManagementToolCoordinator.process(SchemaManagementToolCoordinator.java:85) ~[hibernate-core-5.6.12.Final.jar:5.6.12.Final]
	at org.hibernate.internal.SessionFactoryImpl.<init>(SessionFactoryImpl.java:335) ~[hibernate-core-5.6.12.Final.jar:5.6.12.Final]
	at org.hibernate.boot.internal.SessionFactoryBuilderImpl.build(SessionFactoryBuilderImpl.java:471) ~[hibernate-core-5.6.12.Final.jar:5.6.12.Final]
	at org.hibernate.jpa.boot.internal.EntityManagerFactoryBuilderImpl.build(EntityManagerFactoryBuilderImpl.java:1498) ~[hibernate-core-5.6.12.Final.jar:5.6.12.Final]
	at org.springframework.orm.jpa.vendor.SpringHibernateJpaPersistenceProvider.createContainerEntityManagerFactory(SpringHibernateJpaPersistenceProvider.java:58) ~[spring-orm-5.2.8.RELEASE.jar:5.2.8.RELEASE]
	at org.springframework.orm.jpa.LocalContainerEntityManagerFactoryBean.createNativeEntityManagerFactory(LocalContainerEntityManagerFactoryBean.java:365) ~[spring-orm-5.2.8.RELEASE.jar:5.2.8.RELEASE]
	at org.springframework.orm.jpa.AbstractEntityManagerFactoryBean.buildNativeEntityManagerFactory(AbstractEntityManagerFactoryBean.java:391) ~[spring-orm-5.2.8.RELEASE.jar:5.2.8.RELEASE]
	... 6 more
Caused by: com.google.cloud.spanner.jdbc.JdbcSqlExceptionFactory$JdbcSqlExceptionImpl: FAILED_PRECONDITION: Executing queries is not allowed for DDL batches.
	at com.google.cloud.spanner.jdbc.JdbcSqlExceptionFactory.of(JdbcSqlExceptionFactory.java:221) ~[google-cloud-spanner-jdbc-2.7.9.jar:2.7.9]
	at com.google.cloud.spanner.jdbc.AbstractJdbcStatement.executeQuery(AbstractJdbcStatement.java:198) ~[google-cloud-spanner-jdbc-2.7.9.jar:2.7.9]
	at com.google.cloud.spanner.jdbc.AbstractJdbcStatement.executeQuery(AbstractJdbcStatement.java:179) ~[google-cloud-spanner-jdbc-2.7.9.jar:2.7.9]
	at com.google.cloud.spanner.jdbc.JdbcPreparedStatement.executeQuery(JdbcPreparedStatement.java:69) ~[google-cloud-spanner-jdbc-2.7.9.jar:2.7.9]
	at com.zaxxer.hikari.pool.ProxyPreparedStatement.executeQuery(ProxyPreparedStatement.java:52) ~[HikariCP-3.4.5.jar:?]
	at com.zaxxer.hikari.pool.HikariProxyPreparedStatement.executeQuery(HikariProxyPreparedStatement.java) ~[HikariCP-3.4.5.jar:?]
	at org.hibernate.tool.schema.extract.spi.ExtractionContext.getQueryResults(ExtractionContext.java:49) ~[hibernate-core-5.6.12.Final.jar:5.6.12.Final]
	at org.hibernate.tool.schema.extract.internal.InformationExtractorJdbcDatabaseMetaDataImpl.addColumns(InformationExtractorJdbcDatabaseMetaDataImpl.java:147) ~[hibernate-core-5.6.12.Final.jar:?]
	at org.hibernate.tool.schema.extract.internal.AbstractInformationExtractorImpl.extractTableInformation(AbstractInformationExtractorImpl.java:844) ~[hibernate-core-5.6.12.Final.jar:?]
	at org.hibernate.tool.schema.extract.internal.AbstractInformationExtractorImpl.lambda$getTable$2(AbstractInformationExtractorImpl.java:399) ~[hibernate-core-5.6.12.Final.jar:?]
	at org.hibernate.tool.schema.extract.internal.InformationExtractorJdbcDatabaseMetaDataImpl.processTableResultSet(InformationExtractorJdbcDatabaseMetaDataImpl.java:69) ~[hibernate-core-5.6.12.Final.jar:?]
	at org.hibernate.tool.schema.extract.internal.AbstractInformationExtractorImpl.getTable(AbstractInformationExtractorImpl.java:394) ~[hibernate-core-5.6.12.Final.jar:?]
	at org.hibernate.tool.schema.extract.internal.DatabaseInformationImpl.getTableInformation(DatabaseInformationImpl.java:114) ~[gps.core.jar:?]
	at org.hibernate.tool.schema.extract.internal.DatabaseInformationImpl.locateTableInformation(DatabaseInformationImpl.java:152) ~[gps.core.jar:?]
	at org.hibernate.tool.schema.extract.internal.AbstractInformationExtractorImpl.lambda$getForeignKeys$9(AbstractInformationExtractorImpl.java:1298) ~[hibernate-core-5.6.12.Final.jar:?]
	at org.hibernate.tool.schema.extract.internal.InformationExtractorJdbcDatabaseMetaDataImpl.processImportedKeysResultSet(InformationExtractorJdbcDatabaseMetaDataImpl.java:132) ~[hibernate-core-5.6.12.Final.jar:?]
	at org.hibernate.tool.schema.extract.internal.AbstractInformationExtractorImpl.getForeignKeys(AbstractInformationExtractorImpl.java:1277) ~[hibernate-core-5.6.12.Final.jar:?]
	at org.hibernate.tool.schema.extract.internal.TableInformationImpl.foreignKeys(TableInformationImpl.java:97) ~[hibernate-core-5.6.12.Final.jar:?]
	at org.hibernate.tool.schema.extract.internal.TableInformationImpl.getForeignKeys(TableInformationImpl.java:91) ~[hibernate-core-5.6.12.Final.jar:?]
	at org.hibernate.tool.schema.internal.AbstractSchemaMigrator.equivalentForeignKeyExistsInDatabase(AbstractSchemaMigrator.java:500) ~[hibernate-core-5.6.12.Final.jar:5.6.12.Final]
	at org.hibernate.tool.schema.internal.AbstractSchemaMigrator.checkForExistingForeignKey(AbstractSchemaMigrator.java:485) ~[hibernate-core-5.6.12.Final.jar:5.6.12.Final]
	at org.hibernate.tool.schema.internal.AbstractSchemaMigrator.applyForeignKeys(AbstractSchemaMigrator.java:442) ~[hibernate-core-5.6.12.Final.jar:5.6.12.Final]
	at org.hibernate.tool.schema.internal.AbstractSchemaMigrator.performMigration(AbstractSchemaMigrator.java:263) ~[hibernate-core-5.6.12.Final.jar:5.6.12.Final]
	at org.hibernate.tool.schema.internal.AbstractSchemaMigrator.doMigration(AbstractSchemaMigrator.java:123) ~[hibernate-core-5.6.12.Final.jar:5.6.12.Final]
	at com.google.cloud.spanner.hibernate.schema.SpannerSchemaMigrator.doMigration(SpannerSchemaMigrator.java:57) ~[google-cloud-spanner-hibernate-dialect-1.5.7-SNAPSHOT.jar:1.5.7-SNAPSHOT]
	at org.hibernate.tool.schema.spi.SchemaManagementToolCoordinator.performDatabaseAction(SchemaManagementToolCoordinator.java:196) ~[hibernate-core-5.6.12.Final.jar:5.6.12.Final]
	at org.hibernate.tool.schema.spi.SchemaManagementToolCoordinator.process(SchemaManagementToolCoordinator.java:85) ~[hibernate-core-5.6.12.Final.jar:5.6.12.Final]
	at org.hibernate.internal.SessionFactoryImpl.<init>(SessionFactoryImpl.java:335) ~[hibernate-core-5.6.12.Final.jar:5.6.12.Final]
	at org.hibernate.boot.internal.SessionFactoryBuilderImpl.build(SessionFactoryBuilderImpl.java:471) ~[hibernate-core-5.6.12.Final.jar:5.6.12.Final]
	at org.hibernate.jpa.boot.internal.EntityManagerFactoryBuilderImpl.build(EntityManagerFactoryBuilderImpl.java:1498) ~[hibernate-core-5.6.12.Final.jar:5.6.12.Final]
	at org.springframework.orm.jpa.vendor.SpringHibernateJpaPersistenceProvider.createContainerEntityManagerFactory(SpringHibernateJpaPersistenceProvider.java:58) ~[spring-orm-5.2.8.RELEASE.jar:5.2.8.RELEASE]
	at org.springframework.orm.jpa.LocalContainerEntityManagerFactoryBean.createNativeEntityManagerFactory(LocalContainerEntityManagerFactoryBean.java:365) ~[spring-orm-5.2.8.RELEASE.jar:5.2.8.RELEASE]
	at org.springframework.orm.jpa.AbstractEntityManagerFactoryBean.buildNativeEntityManagerFactory(AbstractEntityManagerFactoryBean.java:391) ~[spring-orm-5.2.8.RELEASE.jar:5.2.8.RELEASE]
	... 6 more
Caused by: com.google.cloud.spanner.SpannerException: FAILED_PRECONDITION: Executing queries is not allowed for DDL batches.
	at com.google.cloud.spanner.SpannerExceptionFactory.newSpannerExceptionPreformatted(SpannerExceptionFactory.java:291) ~[google-cloud-spanner-6.31.2.jar:6.31.2]
	at com.google.cloud.spanner.SpannerExceptionFactory.newSpannerExceptionPreformatted(SpannerExceptionFactory.java:297) ~[google-cloud-spanner-6.31.2.jar:6.31.2]
	at com.google.cloud.spanner.SpannerExceptionFactory.newSpannerException(SpannerExceptionFactory.java:61) ~[google-cloud-spanner-6.31.2.jar:6.31.2]
	at com.google.cloud.spanner.SpannerExceptionFactory.newSpannerException(SpannerExceptionFactory.java:56) ~[google-cloud-spanner-6.31.2.jar:6.31.2]
	at com.google.cloud.spanner.connection.DdlBatch.executeQueryAsync(DdlBatch.java:145) ~[google-cloud-spanner-6.31.2.jar:6.31.2]
	at com.google.cloud.spanner.connection.ConnectionImpl.internalExecuteQuery(ConnectionImpl.java:1148) ~[google-cloud-spanner-6.31.2.jar:6.31.2]
	at com.google.cloud.spanner.connection.ConnectionImpl.parseAndExecuteQuery(ConnectionImpl.java:929) ~[google-cloud-spanner-6.31.2.jar:6.31.2]
	at com.google.cloud.spanner.connection.ConnectionImpl.executeQuery(ConnectionImpl.java:897) ~[google-cloud-spanner-6.31.2.jar:6.31.2]
	at com.google.cloud.spanner.jdbc.AbstractJdbcStatement.executeQuery(AbstractJdbcStatement.java:192) ~[google-cloud-spanner-jdbc-2.7.9.jar:2.7.9]
	at com.google.cloud.spanner.jdbc.AbstractJdbcStatement.executeQuery(AbstractJdbcStatement.java:179) ~[google-cloud-spanner-jdbc-2.7.9.jar:2.7.9]
	at com.google.cloud.spanner.jdbc.JdbcPreparedStatement.executeQuery(JdbcPreparedStatement.java:69) ~[google-cloud-spanner-jdbc-2.7.9.jar:2.7.9]
	at com.zaxxer.hikari.pool.ProxyPreparedStatement.executeQuery(ProxyPreparedStatement.java:52) ~[HikariCP-3.4.5.jar:?]
	at com.zaxxer.hikari.pool.HikariProxyPreparedStatement.executeQuery(HikariProxyPreparedStatement.java) ~[HikariCP-3.4.5.jar:?]
	at org.hibernate.tool.schema.extract.spi.ExtractionContext.getQueryResults(ExtractionContext.java:49) ~[hibernate-core-5.6.12.Final.jar:5.6.12.Final]
	at org.hibernate.tool.schema.extract.internal.InformationExtractorJdbcDatabaseMetaDataImpl.addColumns(InformationExtractorJdbcDatabaseMetaDataImpl.java:147) ~[hibernate-core-5.6.12.Final.jar:?]
	at org.hibernate.tool.schema.extract.internal.AbstractInformationExtractorImpl.extractTableInformation(AbstractInformationExtractorImpl.java:844) ~[hibernate-core-5.6.12.Final.jar:?]
	at org.hibernate.tool.schema.extract.internal.AbstractInformationExtractorImpl.lambda$getTable$2(AbstractInformationExtractorImpl.java:399) ~[hibernate-core-5.6.12.Final.jar:?]
	at org.hibernate.tool.schema.extract.internal.InformationExtractorJdbcDatabaseMetaDataImpl.processTableResultSet(InformationExtractorJdbcDatabaseMetaDataImpl.java:69) ~[hibernate-core-5.6.12.Final.jar:?]
	at org.hibernate.tool.schema.extract.internal.AbstractInformationExtractorImpl.getTable(AbstractInformationExtractorImpl.java:394) ~[hibernate-core-5.6.12.Final.jar:?]
	at org.hibernate.tool.schema.extract.internal.DatabaseInformationImpl.getTableInformation(DatabaseInformationImpl.java:114) ~[gps.core.jar:?]
	at org.hibernate.tool.schema.extract.internal.DatabaseInformationImpl.locateTableInformation(DatabaseInformationImpl.java:152) ~[gps.core.jar:?]
	at org.hibernate.tool.schema.extract.internal.AbstractInformationExtractorImpl.lambda$getForeignKeys$9(AbstractInformationExtractorImpl.java:1298) ~[hibernate-core-5.6.12.Final.jar:?]
	at org.hibernate.tool.schema.extract.internal.InformationExtractorJdbcDatabaseMetaDataImpl.processImportedKeysResultSet(InformationExtractorJdbcDatabaseMetaDataImpl.java:132) ~[hibernate-core-5.6.12.Final.jar:?]
	at org.hibernate.tool.schema.extract.internal.AbstractInformationExtractorImpl.getForeignKeys(AbstractInformationExtractorImpl.java:1277) ~[hibernate-core-5.6.12.Final.jar:?]
	at org.hibernate.tool.schema.extract.internal.TableInformationImpl.foreignKeys(TableInformationImpl.java:97) ~[hibernate-core-5.6.12.Final.jar:?]
	at org.hibernate.tool.schema.extract.internal.TableInformationImpl.getForeignKeys(TableInformationImpl.java:91) ~[hibernate-core-5.6.12.Final.jar:?]
	at org.hibernate.tool.schema.internal.AbstractSchemaMigrator.equivalentForeignKeyExistsInDatabase(AbstractSchemaMigrator.java:500) ~[hibernate-core-5.6.12.Final.jar:5.6.12.Final]
	at org.hibernate.tool.schema.internal.AbstractSchemaMigrator.checkForExistingForeignKey(AbstractSchemaMigrator.java:485) ~[hibernate-core-5.6.12.Final.jar:5.6.12.Final]
	at org.hibernate.tool.schema.internal.AbstractSchemaMigrator.applyForeignKeys(AbstractSchemaMigrator.java:442) ~[hibernate-core-5.6.12.Final.jar:5.6.12.Final]
	at org.hibernate.tool.schema.internal.AbstractSchemaMigrator.performMigration(AbstractSchemaMigrator.java:263) ~[hibernate-core-5.6.12.Final.jar:5.6.12.Final]
	at org.hibernate.tool.schema.internal.AbstractSchemaMigrator.doMigration(AbstractSchemaMigrator.java:123) ~[hibernate-core-5.6.12.Final.jar:5.6.12.Final]
	at com.google.cloud.spanner.hibernate.schema.SpannerSchemaMigrator.doMigration(SpannerSchemaMigrator.java:57) ~[google-cloud-spanner-hibernate-dialect-1.5.7-SNAPSHOT.jar:1.5.7-SNAPSHOT]
	at org.hibernate.tool.schema.spi.SchemaManagementToolCoordinator.performDatabaseAction(SchemaManagementToolCoordinator.java:196) ~[hibernate-core-5.6.12.Final.jar:5.6.12.Final]
	at org.hibernate.tool.schema.spi.SchemaManagementToolCoordinator.process(SchemaManagementToolCoordinator.java:85) ~[hibernate-core-5.6.12.Final.jar:5.6.12.Final]
	at org.hibernate.internal.SessionFactoryImpl.<init>(SessionFactoryImpl.java:335) ~[hibernate-core-5.6.12.Final.jar:5.6.12.Final]
	at org.hibernate.boot.internal.SessionFactoryBuilderImpl.build(SessionFactoryBuilderImpl.java:471) ~[hibernate-core-5.6.12.Final.jar:5.6.12.Final]
	at org.hibernate.jpa.boot.internal.EntityManagerFactoryBuilderImpl.build(EntityManagerFactoryBuilderImpl.java:1498) ~[hibernate-core-5.6.12.Final.jar:5.6.12.Final]
	at org.springframework.orm.jpa.vendor.SpringHibernateJpaPersistenceProvider.createContainerEntityManagerFactory(SpringHibernateJpaPersistenceProvider.java:58) ~[spring-orm-5.2.8.RELEASE.jar:5.2.8.RELEASE]
	at org.springframework.orm.jpa.LocalContainerEntityManagerFactoryBean.createNativeEntityManagerFactory(LocalContainerEntityManagerFactoryBean.java:365) ~[spring-orm-5.2.8.RELEASE.jar:5.2.8.RELEASE]
	at org.springframework.orm.jpa.AbstractEntityManagerFactoryBean.buildNativeEntityManagerFactory(AbstractEntityManagerFactoryBean.java:391) ~[spring-orm-5.2.8.RELEASE.jar:5.2.8.RELEASE]
	... 6 more
```

--- a/google-cloud-spanner-hibernate-dialect/src/main/java/com/google/cloud/spanner/hibernate/schema/SpannerExtractionContext.java
+++ b/google-cloud-spanner-hibernate-dialect/src/main/java/com/google/cloud/spanner/hibernate/schema/SpannerExtractionContext.java
@@ -26,6 +26,7 @@ import org.hibernate.engine.jdbc.env.spi.JdbcEnvironment;
 import org.hibernate.resource.transaction.spi.DdlTransactionIsolator;
 import org.hibernate.service.ServiceRegistry;
 import org.hibernate.tool.schema.internal.exec.ImprovedExtractionContextImpl;
+import org.jboss.logging.Logger;
 
 /**
  * {@link SpannerExtractionContext} uses a separate JDBC connection for extracting metadata from the
@@ -33,6 +34,7 @@ import org.hibernate.tool.schema.internal.exec.ImprovedExtractionContextImpl;
  * SpannerSchemaManagementTool} while that connection is in a DDL batch.
  */
 public class SpannerExtractionContext extends ImprovedExtractionContextImpl {
+  private static final Logger log = Logger.getLogger(SpannerExtractionContext.class);
   private final JdbcConnectionAccess jdbcConnectionAccess;
   private Connection extractionConnection;
 
@@ -59,7 +61,11 @@ public class SpannerExtractionContext extends ImprovedExtractionContextImpl {
       if (extractionConnection == null) {
         extractionConnection = jdbcConnectionAccess.obtainConnection();
       }
-    } catch (SQLException ignore) {
+    } catch (SQLException exception) {
+      log.warn(
+          "An exception was thrown while obtaining a JDBC connection for metadata extraction. "
+              + "Falling back to the default connection used for DDL execution.",
+          exception);
       // Fallback and use the original connection if anything goes wrong.
       extractionConnection = super.getJdbcConnection();
     }
@@ -72,8 +78,11 @@ public class SpannerExtractionContext extends ImprovedExtractionContextImpl {
     if (extractionConnection != null) {
       try {
         extractionConnection.close();
-      } catch (SQLException ignore) {
-        // Nothing we can do, just ignore it.
+      } catch (SQLException exception) {
+        log.warn(
+            "An exception was thrown while closing the JDBC connection "
+                + "that was used for metadata extraction",
+            exception);
       }
     }
   }

--- a/google-cloud-spanner-hibernate-dialect/src/main/java/com/google/cloud/spanner/hibernate/schema/SpannerExtractionContext.java
+++ b/google-cloud-spanner-hibernate-dialect/src/main/java/com/google/cloud/spanner/hibernate/schema/SpannerExtractionContext.java
@@ -1,0 +1,80 @@
+/*
+ * Copyright 2019-2020 Google LLC
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA
+ */
+
+package com.google.cloud.spanner.hibernate.schema;
+
+import java.sql.Connection;
+import java.sql.SQLException;
+import org.hibernate.boot.model.relational.SqlStringGenerationContext;
+import org.hibernate.engine.jdbc.connections.spi.JdbcConnectionAccess;
+import org.hibernate.engine.jdbc.env.spi.JdbcEnvironment;
+import org.hibernate.resource.transaction.spi.DdlTransactionIsolator;
+import org.hibernate.service.ServiceRegistry;
+import org.hibernate.tool.schema.internal.exec.ImprovedExtractionContextImpl;
+
+/**
+ * {@link SpannerExtractionContext} uses a separate JDBC connection for extracting metadata from the
+ * database. This ensures that no queries are executed on the connection used by the {@link
+ * SpannerSchemaManagementTool} while that connection is in a DDL batch.
+ */
+public class SpannerExtractionContext extends ImprovedExtractionContextImpl {
+  private final JdbcConnectionAccess jdbcConnectionAccess;
+  private Connection extractionConnection;
+
+  SpannerExtractionContext(
+      ServiceRegistry serviceRegistry,
+      JdbcEnvironment jdbcEnvironment,
+      SqlStringGenerationContext sqlStringGenerationContext,
+      DdlTransactionIsolator ddlTransactionIsolator,
+      DatabaseObjectAccess databaseObjectAccess) {
+    super(
+        serviceRegistry,
+        jdbcEnvironment,
+        sqlStringGenerationContext,
+        ddlTransactionIsolator,
+        databaseObjectAccess);
+    this.jdbcConnectionAccess = ddlTransactionIsolator.getJdbcContext().getJdbcConnectionAccess();
+  }
+
+  @Override
+  public Connection getJdbcConnection() {
+    // Get a separate JDBC connection for metadata extraction. This makes sure that Hibernate does
+    // not try to extract metadata using a connection that has an active DDL batch.
+    try {
+      if (extractionConnection == null) {
+        extractionConnection = jdbcConnectionAccess.obtainConnection();
+      }
+    } catch (SQLException ignore) {
+      // Fallback and use the original connection if anything goes wrong.
+      extractionConnection = super.getJdbcConnection();
+    }
+    return extractionConnection;
+  }
+
+  @Override
+  public void cleanup() {
+    super.cleanup();
+    if (extractionConnection != null) {
+      try {
+        extractionConnection.close();
+      } catch (SQLException ignore) {
+        // Nothing we can do, just ignore it.
+      }
+    }
+  }
+}

--- a/google-cloud-spanner-hibernate-dialect/src/main/java/com/google/cloud/spanner/hibernate/schema/SpannerSchemaManagementTool.java
+++ b/google-cloud-spanner-hibernate-dialect/src/main/java/com/google/cloud/spanner/hibernate/schema/SpannerSchemaManagementTool.java
@@ -32,10 +32,17 @@ import java.sql.Connection;
 import java.sql.SQLException;
 import java.sql.Statement;
 import java.util.Map;
+import org.hibernate.boot.model.relational.SqlStringGenerationContext;
+import org.hibernate.engine.jdbc.env.spi.JdbcEnvironment;
 import org.hibernate.resource.transaction.spi.DdlTransactionIsolator;
+import org.hibernate.service.ServiceRegistry;
+import org.hibernate.tool.schema.extract.internal.InformationExtractorJdbcDatabaseMetaDataImpl;
+import org.hibernate.tool.schema.extract.spi.ExtractionContext;
+import org.hibernate.tool.schema.extract.spi.InformationExtractor;
 import org.hibernate.tool.schema.internal.HibernateSchemaManagementTool;
 import org.hibernate.tool.schema.internal.exec.JdbcContext;
 import org.hibernate.tool.schema.spi.ExecutionOptions;
+import org.hibernate.tool.schema.spi.ExtractionTool;
 import org.hibernate.tool.schema.spi.SchemaCreator;
 import org.hibernate.tool.schema.spi.SchemaDropper;
 import org.hibernate.tool.schema.spi.SchemaMigrator;
@@ -182,6 +189,43 @@ public class SpannerSchemaManagementTool extends HibernateSchemaManagementTool {
       return new SpannerDdlTransactionIsolator(delegate);
     } catch (Throwable ignore) {
       return delegate;
+    }
+  }
+
+  @Override
+  public ExtractionTool getExtractionTool() {
+    return SpannerExtractionTool.INSTANCE;
+  }
+
+  /**
+   * {@link SpannerExtractionTool} creates an {@link ExtractionContext} that uses a separate JDBC
+   * connection for extracting additional metadata from the database. This prevents queries from
+   * being executed on the connection that is executing the DDL batch for the migration.
+   */
+  private static class SpannerExtractionTool implements ExtractionTool {
+
+    private static final SpannerExtractionTool INSTANCE = new SpannerExtractionTool();
+
+    private SpannerExtractionTool() {}
+
+    @Override
+    public ExtractionContext createExtractionContext(
+        ServiceRegistry serviceRegistry,
+        JdbcEnvironment jdbcEnvironment,
+        SqlStringGenerationContext sqlStringGenerationContext,
+        DdlTransactionIsolator ddlTransactionIsolator,
+        ExtractionContext.DatabaseObjectAccess databaseObjectAccess) {
+      return new SpannerExtractionContext(
+          serviceRegistry,
+          jdbcEnvironment,
+          sqlStringGenerationContext,
+          ddlTransactionIsolator,
+          databaseObjectAccess);
+    }
+
+    @Override
+    public InformationExtractor createInformationExtractor(ExtractionContext extractionContext) {
+      return new InformationExtractorJdbcDatabaseMetaDataImpl(extractionContext);
     }
   }
 

--- a/google-cloud-spanner-hibernate-dialect/src/test/java/com/google/cloud/spanner/hibernate/GeneratedUpdateTableStatementsTests.java
+++ b/google-cloud-spanner-hibernate-dialect/src/test/java/com/google/cloud/spanner/hibernate/GeneratedUpdateTableStatementsTests.java
@@ -19,12 +19,24 @@
 package com.google.cloud.spanner.hibernate;
 
 import static org.assertj.core.api.Assertions.assertThat;
+import static org.junit.Assert.assertEquals;
 
+import com.google.cloud.spanner.hibernate.entities.Customer;
 import com.google.cloud.spanner.hibernate.entities.Employee;
+import com.google.cloud.spanner.hibernate.entities.Invoice;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
 import com.mockrunner.mock.jdbc.JDBCMockObjectFactory;
 import com.mockrunner.mock.jdbc.MockConnection;
+import com.mockrunner.mock.jdbc.MockDriver;
+import com.mockrunner.mock.jdbc.MockResultSet;
+import com.mockrunner.mock.jdbc.MockResultSetMetaData;
+import java.sql.Connection;
 import java.sql.SQLException;
+import java.util.ArrayList;
 import java.util.List;
+import java.util.Map;
+import java.util.Properties;
 import org.hibernate.Session;
 import org.hibernate.boot.Metadata;
 import org.hibernate.boot.MetadataSources;
@@ -33,87 +45,182 @@ import org.hibernate.boot.registry.StandardServiceRegistryBuilder;
 import org.junit.Before;
 import org.junit.Test;
 
-/**
- * Verifies correct DDL generation of update table statements.
- */
+/** Verifies correct DDL generation of update table statements. */
 public class GeneratedUpdateTableStatementsTests {
 
   private StandardServiceRegistry registry;
 
   private JDBCMockObjectFactory jdbcMockObjectFactory;
 
-  private MockConnection mockConnection;
+  private MockConnection defaultConnection;
+  private MockConnection ddlBatchMockConnection;
+  private MockConnection extractorMockConnection;
 
-  /**
-   * Set up the metadata for Hibernate to generate schema statements.
-   */
+  /** Set up the metadata for Hibernate to generate schema statements. */
   @Before
   public void setup() {
-    this.jdbcMockObjectFactory = new JDBCMockObjectFactory();
+    this.jdbcMockObjectFactory =
+        new JDBCMockObjectFactory() {
+          @Override
+          public MockDriver createMockDriver() {
+            return new MockDriver() {
+              final List<Connection> connections = new ArrayList<>();
+              int index = 0;
+
+              @Override
+              public void setupConnection(Connection connection) {
+                connections.add(connection);
+              }
+
+              @Override
+              public Connection connect(String url, Properties info) {
+                Connection connection = connections.get(Math.min(index, connections.size() - 1));
+                index++;
+                return connection;
+              }
+            };
+          }
+        };
     this.jdbcMockObjectFactory.registerMockDriver();
 
-    mockConnection = this.jdbcMockObjectFactory.createMockConnection();
+    defaultConnection = this.jdbcMockObjectFactory.getMockConnection();
+    ddlBatchMockConnection = this.jdbcMockObjectFactory.createMockConnection();
+    extractorMockConnection = this.jdbcMockObjectFactory.createMockConnection();
 
-    this.jdbcMockObjectFactory
-        .getMockDriver()
-        .setupConnection(mockConnection);
+    this.jdbcMockObjectFactory.getMockDriver().setupConnection(ddlBatchMockConnection);
+    this.jdbcMockObjectFactory.getMockDriver().setupConnection(extractorMockConnection);
 
-    this.registry = new StandardServiceRegistryBuilder()
-        .applySetting("hibernate.dialect", SpannerDialect.class.getName())
-        // must NOT set a driver class name so that Hibernate will use java.sql.DriverManager
-        // and discover the only mock driver we have set up.
-        .applySetting("hibernate.connection.url", "unused")
-        .applySetting("hibernate.connection.username", "unused")
-        .applySetting("hibernate.connection.password", "unused")
-        .applySetting("hibernate.hbm2ddl.auto", "update")
-        .build();
+    this.registry =
+        new StandardServiceRegistryBuilder()
+            .applySetting("hibernate.dialect", SpannerDialect.class.getName())
+            // must NOT set a driver class name so that Hibernate will use java.sql.DriverManager
+            // and discover the only mock driver we have set up.
+            .applySetting("hibernate.connection.url", "unused")
+            .applySetting("hibernate.connection.username", "unused")
+            .applySetting("hibernate.connection.password", "unused")
+            .applySetting("hibernate.hbm2ddl.auto", "update")
+            .build();
   }
 
   @Test
   public void testUpdateStatements_createTables() throws SQLException {
     setupTestTables("Hello");
 
-    List<String> sqlStrings = mockConnection.getStatementResultSetHandler().getExecutedStatements();
-    assertThat(sqlStrings).containsExactly(
-        "START BATCH DDL",
-        "create table Employee (id INT64 not null,name STRING(255),manager_id INT64) "
-            + "PRIMARY KEY (id)",
-        "create table hibernate_sequence (next_val INT64) PRIMARY KEY ()",
-        "create index name_index on Employee (name)",
-        "alter table Employee add constraint FKiralam2duuhr33k8a10aoc2t6 "
-            + "foreign key (manager_id) references Employee (id)",
-        "RUN BATCH",
-        "INSERT INTO hibernate_sequence (next_val) VALUES(1)"
-    );
+    List<String> sqlStrings =
+        ddlBatchMockConnection.getStatementResultSetHandler().getExecutedStatements();
+    assertThat(sqlStrings)
+        .containsExactly(
+            "START BATCH DDL",
+            "create table Employee (id INT64 not null,name STRING(255),manager_id INT64) "
+                + "PRIMARY KEY (id)",
+            "create table hibernate_sequence (next_val INT64) PRIMARY KEY ()",
+            "create index name_index on Employee (name)",
+            "alter table Employee add constraint FKiralam2duuhr33k8a10aoc2t6 "
+                + "foreign key (manager_id) references Employee (id)",
+            "RUN BATCH",
+            "INSERT INTO hibernate_sequence (next_val) VALUES(1)");
   }
 
   @Test
   public void testUpdateStatements_alterTables() throws SQLException {
-    setupTestTables("Employee");
-    List<String> sqlStrings = mockConnection.getStatementResultSetHandler().getExecutedStatements();
+    setupTestTables(ImmutableMap.of("Employee", ImmutableList.of("id", "name", "manager_id")));
+    List<String> sqlStrings =
+        ddlBatchMockConnection.getStatementResultSetHandler().getExecutedStatements();
     // The "alter table Employee ADD COLUMN" statements are no longer included in the generated
     // schema when updating an existing table. This schema change was introduced from Hibernate
     // 5.5.2 onwards.
-    assertThat(sqlStrings).containsExactly(
-        "START BATCH DDL",
-        "create table hibernate_sequence (next_val INT64) PRIMARY KEY ()",
-        "create index name_index on Employee (name)",
-        "alter table Employee add constraint FKiralam2duuhr33k8a10aoc2t6 "
-            + "foreign key (manager_id) references Employee (id)",
-        "RUN BATCH",
-        "INSERT INTO hibernate_sequence (next_val) VALUES(1)"
-    );
+    assertThat(sqlStrings)
+        .containsExactly(
+            "START BATCH DDL",
+            "create table hibernate_sequence (next_val INT64) PRIMARY KEY ()",
+            "create index name_index on Employee (name)",
+            "alter table Employee add constraint FKiralam2duuhr33k8a10aoc2t6 "
+                + "foreign key (manager_id) references Employee (id)",
+            "RUN BATCH",
+            "INSERT INTO hibernate_sequence (next_val) VALUES(1)");
   }
 
-  /**
-   * Sets up which pre-existing tables that Hibernate sees.
-   */
+  @Test
+  public void testUpdateStatements_withExistingTables_withForeignKeys() throws SQLException {
+    setupTestTablesWithForeignKeys(
+        ImmutableMap.of(
+            "Customer", ImmutableList.of("customerId", "name"),
+            "customerId", ImmutableList.of("next_val"),
+            "Invoice", ImmutableList.of("invoiceId", "number", "customer_customerId"),
+            "invoiceId", ImmutableList.of("next_val")));
+
+    List<String> sqlStrings =
+        ddlBatchMockConnection.getStatementResultSetHandler().getExecutedStatements();
+    assertThat(sqlStrings).containsExactly("START BATCH DDL", "RUN BATCH");
+    assertEquals(
+        ImmutableList.of(),
+        ddlBatchMockConnection.getPreparedStatementResultSetHandler().getExecutedStatements());
+
+    assertEquals(
+        ImmutableList.of(),
+        extractorMockConnection.getStatementResultSetHandler().getExecutedStatements());
+    assertEquals(
+        ImmutableList.of("select * from `Customer` where 1=0"),
+        extractorMockConnection.getPreparedStatementResultSetHandler().getExecutedStatements());
+  }
+
+  /** Sets up which pre-existing tables that Hibernate sees. */
   private void setupTestTables(String... tables) throws SQLException {
-    mockConnection.setMetaData(MockJdbcUtils.metaDataBuilder().setTables(tables).build());
+    defaultConnection.setMetaData(MockJdbcUtils.metaDataBuilder().setTables(tables).build());
+    extractorMockConnection.setMetaData(MockJdbcUtils.metaDataBuilder().setTables(tables).build());
+
+    Metadata metadata =
+        new MetadataSources(this.registry).addAnnotatedClass(Employee.class).buildMetadata();
+
+    Session session = metadata.buildSessionFactory().openSession();
+    session.beginTransaction();
+    session.close();
+  }
+
+  private void setupTestTables(Map<String, List<String>> tablesAndColumns) throws SQLException {
+    defaultConnection.setMetaData(
+        MockJdbcUtils.metaDataBuilder().setTables(tablesAndColumns).build());
+    extractorMockConnection.setMetaData(
+        MockJdbcUtils.metaDataBuilder().setTables(tablesAndColumns).build());
+
+    Metadata metadata =
+        new MetadataSources(this.registry).addAnnotatedClass(Employee.class).buildMetadata();
+
+    Session session = metadata.buildSessionFactory().openSession();
+    session.beginTransaction();
+    session.close();
+  }
+
+  private void setupTestTablesWithForeignKeys(Map<String, List<String>> tables)
+      throws SQLException {
+    defaultConnection.setMetaData(
+        MockJdbcUtils.metaDataBuilder()
+            .setTables(tables)
+            .setImportedKeys(
+                "Customer", "customerId", "Invoice", "customer_customerId", "fk_invoice_customer")
+            .build());
+    extractorMockConnection.setMetaData(
+        MockJdbcUtils.metaDataBuilder()
+            .setTables(tables)
+            .setImportedKeys(
+                "Customer", "customerId", "Invoice", "customer_customerId", "fk_invoice_customer")
+            .build());
+    MockResultSetMetaData mockResultSetMetaData = new MockResultSetMetaData();
+    mockResultSetMetaData.setColumnCount(2);
+    mockResultSetMetaData.setColumnName(1, "customerId");
+    mockResultSetMetaData.setColumnTypeName(1, "INT64");
+    mockResultSetMetaData.setColumnName(2, "name");
+    mockResultSetMetaData.setColumnTypeName(2, "STRING");
+    MockResultSet resultSet = new MockResultSet("invoice_columns");
+    resultSet.setResultSetMetaData(mockResultSetMetaData);
+    extractorMockConnection
+        .getPreparedStatementResultSetHandler()
+        .prepareResultSet("select * from `Customer` where 1=0", resultSet);
 
     Metadata metadata =
         new MetadataSources(this.registry)
-            .addAnnotatedClass(Employee.class)
+            .addAnnotatedClass(Customer.class)
+            .addAnnotatedClass(Invoice.class)
             .buildMetadata();
 
     Session session = metadata.buildSessionFactory().openSession();

--- a/google-cloud-spanner-hibernate-dialect/src/test/java/com/google/cloud/spanner/hibernate/MockJdbcUtils.java
+++ b/google-cloud-spanner-hibernate-dialect/src/test/java/com/google/cloud/spanner/hibernate/MockJdbcUtils.java
@@ -18,13 +18,18 @@
 
 package com.google.cloud.spanner.hibernate;
 
+import com.google.common.collect.ImmutableList;
 import com.mockrunner.mock.jdbc.MockDatabaseMetaData;
 import com.mockrunner.mock.jdbc.MockResultSet;
 import java.sql.ResultSet;
 import java.sql.SQLException;
 import java.sql.Types;
 import java.util.Arrays;
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
 import java.util.UUID;
+import java.util.stream.Collectors;
 
 /**
  * Helper class to building mock objects for the mock JDBC driver.
@@ -41,6 +46,13 @@ public class MockJdbcUtils {
       "TYPE_NAME", "COLUMN_SIZE", "DECIMAL_DIGITS", "IS_NULLABLE",
   };
 
+  private static final String[] IMPORTED_KEY_COLUMNS = new String[]{
+      "PKTABLE_CAT", "PKTABLE_SCHEM", "PKTABLE_NAME", "PKCOLUMN_NAME",
+      "FKTABLE_CAT", "FKTABLE_SCHEM", "FKTABLE_NAME", "FKCOLUMN_NAME",
+      "KEY_SEQ", "UPDATE_RULE", "DELETE_RULE", "FK_NAME", "PK_NAME",
+      "DEFERRABILITY"
+  };
+
   /**
    * Creates the metadata object read by Hibernate to determine which tables already exist.
    */
@@ -51,20 +63,22 @@ public class MockJdbcUtils {
   /**
    * Constructs a mock Column metadata {@link ResultSet} describing fake columns of the tables.
    */
-  private static ResultSet createColumnMetadataResultSet(String... tableNames) {
+  private static ResultSet createColumnMetadataResultSet(Map<String, List<String>> columns) {
     MockResultSet mockResultSet = initResultSet(COLUMN_METADATA_LABELS);
 
-    for (int i = 0; i < tableNames.length; i++) {
-      Object[] row = new Object[COLUMN_METADATA_LABELS.length];
-      Arrays.fill(row, "");
-      row[2] = tableNames[i];
-      row[3] = "column";
-      row[4] = Types.VARCHAR;
-      row[5] = "string(255)";
-      row[6] = 255;
-      row[7] = 0;
-      row[8] = 0;
-      mockResultSet.addRow(row);
+    for (Entry<String, List<String>> entry : columns.entrySet()) {
+      for (String column : entry.getValue()) {
+        Object[] row = new Object[COLUMN_METADATA_LABELS.length];
+        Arrays.fill(row, "");
+        row[2] = entry.getKey();
+        row[3] = column;
+        row[4] = Types.VARCHAR;
+        row[5] = "string(255)";
+        row[6] = 255;
+        row[7] = 0;
+        row[8] = 0;
+        mockResultSet.addRow(row);
+      }
     }
 
     return mockResultSet;
@@ -103,6 +117,26 @@ public class MockJdbcUtils {
 
   }
 
+  private static ResultSet createImportedKeysResultSet(
+      String pkTable, String pkColumn, String fkTable, String fkColumn, String fkName) {
+
+    String[] row = new String[IMPORTED_KEY_COLUMNS.length];
+    Arrays.fill(row, "");
+    row[0] = ""; // pk catalog
+    row[1] = ""; // fk catalog
+    row[2] = pkTable;
+    row[3] = pkColumn;
+    row[4] = ""; // fk catalog
+    row[5] = ""; // fk schema
+    row[6] = fkTable;
+    row[7] = fkColumn;
+    row[11] = fkName;
+    MockResultSet mockResultSet = initResultSet(IMPORTED_KEY_COLUMNS);
+    mockResultSet.addRow(row);
+
+    return mockResultSet;
+  }
+
   private static ResettingMockResultSet initResultSet(String... columnLabels) {
     ResettingMockResultSet mockResultSet =
         new ResettingMockResultSet(UUID.randomUUID().toString());
@@ -124,12 +158,23 @@ public class MockJdbcUtils {
     private ResultSet importedKeys = new MockResultSet(UUID.randomUUID().toString());
     private ResultSet exportedKeys = new MockResultSet(UUID.randomUUID().toString());
 
-    /**
-     * Sets which tables are present in the Spanner database.
-     */
+    /** Sets which tables are present in the Spanner database. */
     public MockDatabaseMetaDataBuilder setTables(String... tables) {
       this.tables = createTableMetadataResultSet(tables);
-      this.columns = createColumnMetadataResultSet(tables);
+      this.columns =
+          createColumnMetadataResultSet(
+              Arrays.stream(tables)
+                  .map(t -> new String[] {t, "column"})
+                  .collect(
+                      Collectors.<String[], String, List<String>>toMap(
+                          c -> c[0], c -> ImmutableList.of(c[1]))));
+      return this;
+    }
+
+    /** Sets which tables with which columns are present in the Spanner database. */
+    public MockDatabaseMetaDataBuilder setTables(Map<String, List<String>> tablesAndColumns) {
+      this.tables = createTableMetadataResultSet(tablesAndColumns.keySet().toArray(new String[0]));
+      this.columns = createColumnMetadataResultSet(tablesAndColumns);
       return this;
     }
 
@@ -141,12 +186,25 @@ public class MockJdbcUtils {
       return this;
     }
 
+    /** Sets the imported keys that should be returned. */
+    public MockDatabaseMetaDataBuilder setImportedKeys(
+        String pkTable, String pkColumn, String fkTable, String fkColumn, String fkName) {
+      this.importedKeys = createImportedKeysResultSet(pkTable, pkColumn, fkTable, fkColumn, fkName);
+      return this;
+    }
+
     /**
      * Builds the {@link MockDatabaseMetaData} object which is used by Hibernate to get information
      * about the Spanner Database.
      */
     public MockDatabaseMetaData build() {
       MockDatabaseMetaData mockDatabaseMetaData = new MockDatabaseMetaData();
+      mockDatabaseMetaData.setStoresLowerCaseIdentifiers(false);
+      mockDatabaseMetaData.setStoresUpperCaseIdentifiers(false);
+      mockDatabaseMetaData.setStoresMixedCaseIdentifiers(true);
+      mockDatabaseMetaData.setStoresLowerCaseQuotedIdentifiers(false);
+      mockDatabaseMetaData.setStoresUpperCaseQuotedIdentifiers(false);
+      mockDatabaseMetaData.setStoresMixedCaseIdentifiers(true);
       mockDatabaseMetaData.setColumns(columns);
       mockDatabaseMetaData.setTables(tables);
       mockDatabaseMetaData.setIndexInfo(indexInfo);

--- a/google-cloud-spanner-hibernate-dialect/src/test/java/com/google/cloud/spanner/hibernate/entities/Invoice.java
+++ b/google-cloud-spanner-hibernate-dialect/src/test/java/com/google/cloud/spanner/hibernate/entities/Invoice.java
@@ -1,0 +1,59 @@
+/*
+ * Copyright 2019-2020 Google LLC
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA
+ */
+
+package com.google.cloud.spanner.hibernate.entities;
+
+import javax.persistence.Column;
+import javax.persistence.Entity;
+import javax.persistence.ForeignKey;
+import javax.persistence.GeneratedValue;
+import javax.persistence.GenerationType;
+import javax.persistence.Id;
+import javax.persistence.JoinColumn;
+import javax.persistence.ManyToOne;
+import org.hibernate.annotations.GenericGenerator;
+import org.hibernate.annotations.Parameter;
+import org.hibernate.id.enhanced.SequenceStyleGenerator;
+
+/**
+ * A test entity that has a many-to-one relationship.
+ *
+ * @author loite
+ */
+@Entity
+public class Invoice {
+  @Id
+  @GeneratedValue(strategy = GenerationType.SEQUENCE, generator = "invoiceId")
+  @GenericGenerator(
+      name = "invoiceId",
+      strategy = "com.google.cloud.spanner.hibernate.BitReversedSequenceStyleGenerator",
+      parameters = {
+          @Parameter(name = SequenceStyleGenerator.INCREMENT_PARAM, value = "1000"),
+          @Parameter(name = SequenceStyleGenerator.SEQUENCE_PARAM, value = "invoiceId"),
+          @Parameter(name = SequenceStyleGenerator.INITIAL_PARAM, value = "1")
+      })
+  @Column(nullable = false)
+  private Long invoiceId;
+
+  private String number;
+
+  @ManyToOne
+  @JoinColumn(foreignKey = @ForeignKey(name = "fk_invoice_customer"))
+  private Customer customer;
+
+}

--- a/google-cloud-spanner-hibernate-dialect/src/test/java/com/google/cloud/spanner/hibernate/schema/SpannerExtractionContextTest.java
+++ b/google-cloud-spanner-hibernate-dialect/src/test/java/com/google/cloud/spanner/hibernate/schema/SpannerExtractionContextTest.java
@@ -1,0 +1,157 @@
+/*
+ * Copyright 2019-2020 Google LLC
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA
+ */
+
+package com.google.cloud.spanner.hibernate.schema;
+
+import static org.junit.Assert.assertSame;
+import static org.mockito.Mockito.doThrow;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.never;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
+
+import java.sql.Connection;
+import java.sql.SQLException;
+import org.hibernate.boot.model.relational.SqlStringGenerationContext;
+import org.hibernate.engine.jdbc.connections.spi.JdbcConnectionAccess;
+import org.hibernate.engine.jdbc.env.spi.JdbcEnvironment;
+import org.hibernate.resource.transaction.spi.DdlTransactionIsolator;
+import org.hibernate.service.ServiceRegistry;
+import org.hibernate.tool.schema.extract.spi.ExtractionContext.DatabaseObjectAccess;
+import org.hibernate.tool.schema.internal.exec.JdbcContext;
+import org.junit.Test;
+
+/** Tests for {@link SpannerExtractionContext}. */
+public class SpannerExtractionContextTest {
+
+  @Test
+  public void testGetJdbcConnection_obtainsNewConnection() throws SQLException {
+    DdlTransactionIsolator ddlTransactionIsolator = mock(DdlTransactionIsolator.class);
+    JdbcContext jdbcContext = mock(JdbcContext.class);
+    JdbcConnectionAccess jdbcConnectionAccess = mock(JdbcConnectionAccess.class);
+    Connection connection = mock(Connection.class);
+    when(ddlTransactionIsolator.getJdbcContext()).thenReturn(jdbcContext);
+    when(jdbcContext.getJdbcConnectionAccess()).thenReturn(jdbcConnectionAccess);
+
+    when(jdbcConnectionAccess.obtainConnection()).thenReturn(connection);
+    SpannerExtractionContext context =
+        new SpannerExtractionContext(
+            mock(ServiceRegistry.class),
+            mock(JdbcEnvironment.class),
+            mock(SqlStringGenerationContext.class),
+            ddlTransactionIsolator,
+            mock(DatabaseObjectAccess.class));
+
+    assertSame(connection, context.getJdbcConnection());
+  }
+
+  @Test
+  public void testGetJdbcConnection_fallbacksToDdlConnectionOnException() throws SQLException {
+    DdlTransactionIsolator ddlTransactionIsolator = mock(DdlTransactionIsolator.class);
+    JdbcContext jdbcContext = mock(JdbcContext.class);
+    JdbcConnectionAccess jdbcConnectionAccess = mock(JdbcConnectionAccess.class);
+    Connection ddlConnection = mock(Connection.class);
+    when(ddlTransactionIsolator.getJdbcContext()).thenReturn(jdbcContext);
+    when(jdbcContext.getJdbcConnectionAccess()).thenReturn(jdbcConnectionAccess);
+    when(ddlTransactionIsolator.getIsolatedConnection()).thenReturn(ddlConnection);
+
+    when(jdbcConnectionAccess.obtainConnection()).thenThrow(new SQLException("test"));
+    SpannerExtractionContext context =
+        new SpannerExtractionContext(
+            mock(ServiceRegistry.class),
+            mock(JdbcEnvironment.class),
+            mock(SqlStringGenerationContext.class),
+            ddlTransactionIsolator,
+            mock(DatabaseObjectAccess.class));
+
+    assertSame(ddlConnection, context.getJdbcConnection());
+  }
+
+  @Test
+  public void testCleanupClosesConnection() throws SQLException {
+    DdlTransactionIsolator ddlTransactionIsolator = mock(DdlTransactionIsolator.class);
+    JdbcContext jdbcContext = mock(JdbcContext.class);
+    JdbcConnectionAccess jdbcConnectionAccess = mock(JdbcConnectionAccess.class);
+    Connection connection = mock(Connection.class);
+    when(ddlTransactionIsolator.getJdbcContext()).thenReturn(jdbcContext);
+    when(jdbcContext.getJdbcConnectionAccess()).thenReturn(jdbcConnectionAccess);
+
+    when(jdbcConnectionAccess.obtainConnection()).thenReturn(connection);
+    SpannerExtractionContext context =
+        new SpannerExtractionContext(
+            mock(ServiceRegistry.class),
+            mock(JdbcEnvironment.class),
+            mock(SqlStringGenerationContext.class),
+            ddlTransactionIsolator,
+            mock(DatabaseObjectAccess.class));
+
+    // Ensure that we initialize a connection.
+    context.getJdbcConnection();
+    context.cleanup();
+
+    verify(connection).close();
+  }
+
+  @Test
+  public void testCleanupIsNoOpWithoutConnection() throws SQLException {
+    DdlTransactionIsolator ddlTransactionIsolator = mock(DdlTransactionIsolator.class);
+    JdbcContext jdbcContext = mock(JdbcContext.class);
+    JdbcConnectionAccess jdbcConnectionAccess = mock(JdbcConnectionAccess.class);
+    Connection connection = mock(Connection.class);
+    when(ddlTransactionIsolator.getJdbcContext()).thenReturn(jdbcContext);
+    when(jdbcContext.getJdbcConnectionAccess()).thenReturn(jdbcConnectionAccess);
+
+    when(jdbcConnectionAccess.obtainConnection()).thenReturn(connection);
+    SpannerExtractionContext context =
+        new SpannerExtractionContext(
+            mock(ServiceRegistry.class),
+            mock(JdbcEnvironment.class),
+            mock(SqlStringGenerationContext.class),
+            ddlTransactionIsolator,
+            mock(DatabaseObjectAccess.class));
+
+    // Cleaning up without ever having used the context should be no-op.
+    context.cleanup();
+
+    verify(connection, never()).close();
+  }
+
+  @Test
+  public void testCleanupIgnoresException() throws SQLException {
+    DdlTransactionIsolator ddlTransactionIsolator = mock(DdlTransactionIsolator.class);
+    JdbcContext jdbcContext = mock(JdbcContext.class);
+    JdbcConnectionAccess jdbcConnectionAccess = mock(JdbcConnectionAccess.class);
+    Connection connection = mock(Connection.class);
+    when(ddlTransactionIsolator.getJdbcContext()).thenReturn(jdbcContext);
+    when(jdbcContext.getJdbcConnectionAccess()).thenReturn(jdbcConnectionAccess);
+    when(jdbcConnectionAccess.obtainConnection()).thenReturn(connection);
+    doThrow(new SQLException("test")).when(connection).close();
+    SpannerExtractionContext context =
+        new SpannerExtractionContext(
+            mock(ServiceRegistry.class),
+            mock(JdbcEnvironment.class),
+            mock(SqlStringGenerationContext.class),
+            ddlTransactionIsolator,
+            mock(DatabaseObjectAccess.class));
+
+    context.getJdbcConnection();
+    // Cleanup should ignore any errors.
+    context.cleanup();
+    verify(connection).close();
+  }
+}

